### PR TITLE
GNINA molecular docking with CNNs

### PR DESCRIPTION
**Pull request by ncfrey**
## Description

This PR address #2372, adding support for the new stable release of [GNINA](https://github.com/gnina/gnina) through DeepChem's docking API.  GNINA provides CNN docking scores that outperform AutoDock Vina for pose generation and virtual screening. The ability to run GNINA on GPUs and train new CNN scoring models also brings the docking module more in line with DeepChem's core deep learning focus. 

Thanks to @dkoes, there is a binary distribution with the [ v1.0 release](https://github.com/gnina/gnina/releases/tag/v1.0) that makes installation of GNINA very quick on Colab and most local machines with CUDA>=10.1.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

CCing @dkoes @rbharath @peastman who may be interested

### CODE REVIEW

**Comment by peastman** [comment]
If this feature is only supported on Linux, we should say that in the documentation, and also make it skip the tests on other platforms.  You have it skipping one of the tests on Windows, but not the other one, and it doesn't skip either of them on MacOS.

Is it possible to support this on other platforms?  They only have a precompiled binary for Linux, but they recommend people compile it themselves rather than use that version, and I couldn't find any mention of whether it works on Mac or Windows.  Is there a way we can let people use a binary they compiled themselves?

If vina_utils.py is going to include functions for gnina, perhaps we should rename it to docking_utils.py?

**Comment by dkoes** [comment]
I would need someone to contribute a Windows build.  Building on Mac requires CUDA, which isn't impossible but you'll never actually have GPU acceleration.

File: deepchem/dock/pose_generation.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from deepchem.utils.typing import RDKitMol
 from deepchem.utils.geometry_utils import compute_centroid, compute_protein_range
 from deepchem.utils.rdkit_utils import load_molecule, write_molecule
-from deepchem.utils.vina_utils import load_docked_ligands, write_vina_conf
+from deepchem.utils.vina_utils import load_docked_ligands, write_vina_conf, write_gnina_conf, read_gnina_log
```
Comment (rbharath) [inline deepchem/dock/pose_generation.py:19]: +1 for @peastman's suggestion to rename this `deepchem.utils.docking_utils`.

On a more general note, we should figure out our policy with API stability on utils files. To be safe, we could make the shift backwards compatible by having wrappers in `vina_utils` that emit deprecation warnings.

File: deepchem/dock/pose_generation.py [hunk 2]
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
     raise NotImplementedError
 
 
+class GninaPoseGenerator(PoseGenerator):
+  """Use GNINA to generate binding poses.
+
+  This class uses GNINA (a deep learning framework for molecular
```
Comment (rbharath) [inline deepchem/dock/pose_generation.py:85]: This is excellent documentation!

File: deepchem/dock/pose_generation.py [hunk 3]
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
...
+
+  References
+  ----------
+  .. [1] M Ragoza, J Hochuli, E Idrobo, J Sunseri, DR Koes.
+  "Protein–Ligand Scoring with Convolutional Neural Networks."
+  Journal of chemical information and modeling (2017).
+
+  Notes
+  -----
+  * GNINA requires CUDA >= 10.1 for fast CNN scoring.
```
Comment (rbharath) [inline deepchem/dock/pose_generation.py:107]: Can we add a note here that Gnina currently only works on linux?

File: deepchem/dock/pose_generation.py [hunk 4]
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
...
+  papers: https://github.com/gnina/gnina#citation
+  The primary citation for GNINA is [1]_.
+
+  References
+  ----------
+  .. [1] M Ragoza, J Hochuli, E Idrobo, J Sunseri, DR Koes.
+  "Protein–Ligand Scoring with Convolutional Neural Networks."
+  Journal of chemical information and modeling (2017).
+
+  Notes
```
Comment (rbharath) [inline deepchem/dock/pose_generation.py:105]: Can you change this to "Note" instead? That makes the readthedocs render a nice note blurb

File: deepchem/dock/pose_generation.py [hunk 5]
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
...
+
+    data_dir = get_data_dir()
+    if platform.system() == 'Linux':
+      url = "https://github.com/gnina/gnina/releases/download/v1.0/gnina"
+      filename = 'gnina'
+      self.gnina_dir = data_dir
+      self.gnina_cmd = os.path.join(self.gnina_dir, filename)
+    else:
+      raise ValueError(
+          "Unknown operating system. Try using a cloud platform to run this code instead."
```
Comment (rbharath) [inline deepchem/dock/pose_generation.py:125]: Perhaps "Gnina currently only runs on Linux. Try using a cloud platform to run this code instead."

File: deepchem/utils/test/data/gnina_log.txt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+              _
```
Comment (rbharath) [inline deepchem/utils/test/data/gnina_log.txt:1]: As a quick question for @dkoes, can we assume that this log file format is stable? The PR assumes that the log file will follow this set structure to read/write log files gnina generates.

**Review by rbharath — COMMENTED** [review]
This is excellent! I have a few minor comments below

File: deepchem/utils/test/data/gnina_log.txt (hunk 6 above)
Comment (dkoes) [reply deepchem/utils/test/data/gnina_log.txt:1]: I can't guarantee it in perpetuity, but there are no changes planned that would invalidate the processing.  If you want a more robust solution, extract the scores from the output sdf.

**Comment by ncfrey** [comment]
I addressed the comments above, with the biggest changes being skipping tests that aren't on Linux and deprecating `vina_utils.` To retain compatibility with Python3.6, I used a `@deprecated` decorator function. Happy to use a different implementation if there is a preferred way to deprecate the module.

**Comment by peastman** [comment]
I think we want to just rename vina_utils.py to docking_utils.py.  Keeping around duplicate copies of the functions is a great opportunity for bugs!

File: deepchem/utils/vina_utils.py [hunk 7]
```diff
@@ -56,84 +78,7 @@ def write_vina_conf(protein_filename: str,
...
-      mode = line.split()
-      score = [float(x) for x in mode[1:]]
-      scores.append(score)
-
-  scores = np.array(scores)
-  return scores
-
-
+@deprecated
 def load_docked_ligands(
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:82]: Perhaps we could import the functions from `docking_utils.py` and then have only deprecation wrappers in `vina_utils.py` so we don't have duplicate code?

**Review by rbharath — COMMENTED** [review]
Looking good overall! +1 to @peastman's comment and a suggestion below on how to implement it

File: deepchem/utils/vina_utils.py (hunk 7 above)
Comment (ncfrey) [reply deepchem/utils/vina_utils.py:82]: Got it, will do!

**Comment by ncfrey** [comment]
I removed the duplicate code as suggested by @peastman, following @rbharath's suggestion to wrap the functions in `vina_utils`.

**Review by rbharath — APPROVED** [review]
This PR is looking good! I think this is ready to merge in on my end :)

 Will leave open in case for a bit in case other folks still have comments

**Comment by ncfrey** [comment]
The same pymatgen-related errors are showing up here, as in #2386. I think this is ok to merge, but I'll open a new issue to note that these errors need to be fixed.

**Comment by ncfrey** [comment]
Other than docs, this PR is now passing all tests and should be ready to merge.

**Comment by rbharath** [comment]
Great! Going to merge this in

**Pull request merged**